### PR TITLE
Use at::Tensor based autograd Variable

### PR DESCRIPTION
**Pull request by colesbury**
Variable is now a subclass of `at::Tensor` backed by a `VariableTensor* pImpl`. The implementation of the ATen functions is defined in the auto-generated VariableType.h/cpp file.

Currently, only functions which delegate to the base type (such as `sizes()` and `isCuda()`) are implemented on VariableType. Differentiable ops like `add()` and `mul()` will be added in a subsequent PR.

### CODE REVIEW

File: torch/lib/ATen/templates/Tensor.h [hunk 1]
```diff
@@ -126,7 +126,7 @@ struct Tensor {
   template<typename T>
   T * data() const;
 
-  void * unsafeGetTH(bool retain) {
+  void * unsafeGetTH(bool retain) const {
```
Comment (colesbury) [inline torch/lib/ATen/templates/Tensor.h]: This change is in https://github.com/zdevito/ATen/pull/57. I will rebase on top of that once it lands.

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 2]
```diff
@@ -6,80 +6,55 @@
...
-    // give him the new Value, we just assume that they know these attributes
-    // are changing when using higher order graphs.
-    if (!var_grad->is_volatile || !new_grad->is_volatile) return variable_list();
-    acc_inplace(var_grad, new_grad);
-    return variable_list();
-  }
 
-  if (var->grad_fn)
+  if (!grads[0].defined())
+    return {};
```
Comment (apaszke) [inline torch/csrc/autograd/functions/accumulate_grad.cpp]: We should call hooks if there are any

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 3]
```diff
@@ -6,80 +6,55 @@
...
+  if (!variable.defined())
+    throw std::logic_error("leaf variable was freed");
```
Comment (apaszke) [inline torch/csrc/autograd/functions/accumulate_grad.cpp]: So? Leaf Variables can be freed and it's not an error so let's not imply this. Mention that it has to do with the fact that it was still used in the graph.

Also, it's a condition that we handle now, so it's a breaking change. Not sure how important this is

File: test/test_autograd.py [hunk 4]
```diff
@@ -141,12 +141,10 @@ def test_accumulate_grad(self):
             y.backward(go1, retain_graph=True)
             x_grad = x.grad
             x_grad_clone = x.grad.data.clone()
-
-            del x
             y.backward(go2)
 
             # That's the only case when we can accumulate in-place
-            if start_volatile and end_volatile:
+            if start_volatile:
```
Comment (apaszke) [inline test/test_autograd.py]: I think it would make sense to make it work so that the grad is accumulated in-place only if both start and end are volatile (otherwise we should promote a volatile grad to a non-volatile one)

**Review by apaszke — COMMENTED** [review]
Can you please rename `VariableTensor` to `VariableImpl`? In no way is `VariableTensor` a tensor. It's just a struct that holds a few fields, and it was super confusing.


~~Also, I think adding a strong reference in AccumulateGrad can cause reference cycles in the C++ side. Consider this:~~
```python
x = Variable(torch.ones(1), requires_grad=True)
y = x * 2 # Creates AccumulateGrad for x
x.requires_grad = False # Valid, because x is a leaf
x[0] = y
``` 

~~Now, we have a cycle in the graph: `x -> IndexBackward -> MulBackward -> AccumulateGrad (for x) -> x`~~

EDIT: nevermind. We detach `AccumulateGrad` nodes when moving Variables in the graph

File: torch/csrc/autograd/input_buffer.cpp [hunk 5]
```diff
@@ -9,15 +9,13 @@ InputBuffer::InputBuffer(size_t size)
...
+  if (!var.defined()) {
     return;
   }
   auto& item = buffer[pos];
-  auto& saved_var_ptr = item.first;
-  if (!saved_var_ptr) {
-    auto version = **var->version_counter;
-    buffer[pos] = std::make_pair<>(std::move(var), version);
+  if (!item.first.defined()) {
+    buffer[pos] = std::make_pair<>(std::move(var), var.current_version());
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/input_buffer.cpp]: Oops. This shouldn't be a const reference.

File: torch/csrc/autograd/python_engine.cpp [hunk 6]
```diff
@@ -10,6 +10,7 @@
 #include <unordered_set>
 
 using namespace torch::autograd;
+using at::Tensor;
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/python_engine.cpp]: Removed

File: torch/csrc/autograd/python_variable.cpp [hunk 7]
```diff
@@ -11,19 +11,23 @@
...
+static PyObject* THPVariable_NewWithVar(PyTypeObject* type, const Variable& var)
 {
   PyObject* obj = type->tp_alloc(type, 0);
   if (obj) {
     auto v = (THPVariable*) obj;
-    new (&v->cdata) std::shared_ptr<Variable>(std::move(var));
-    if (auto fn = dynamic_cast<PyFunction*>(v->cdata->grad_fn.get())) {
+    v->cdata = var.get();
+    v->cdata->retain();
+    v->cdata->pyobj = obj;
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/python_variable.cpp]: I've added a comment. We've had the invariant for a while that every Variable has only one Python wrapper associated with it (or no wrapper).

File: torch/csrc/autograd/functions/accumulate_grad.cpp (hunk 3 above)
Comment (colesbury) [reply torch/csrc/autograd/functions/accumulate_grad.cpp]: We can probably remove the check. AccumulateGrad has a strong pointer to `variable` so it won't get freed unless someone explicitly calls `AccumulateGrad.variable.reset()` or something similar.

File: torch/csrc/autograd/functions/accumulate_grad.cpp (hunk 2 above)
Comment (colesbury) [reply torch/csrc/autograd/functions/accumulate_grad.cpp]: We didn't before. If you want to change that behavior we should do that in a separate commit, since it may introduce potential bugs.

File: test/test_autograd.py (hunk 4 above)
Comment (colesbury) [reply test/test_autograd.py]: This matches the old behavior for most scenarios (where the parent variable `x` doesn't go out of scope).

File: torch/csrc/autograd/functions/accumulate_grad.cpp (hunk 3 above)
Comment (apaszke) [reply torch/csrc/autograd/functions/accumulate_grad.cpp]: Yes, that sounds good. The check is a bit confusing, but a comment would be nice

File: test/test_autograd.py (hunk 4 above)
Comment (apaszke) [reply test/test_autograd.py]: I know, but as we discussed the old behavior might not make sense. Anyway, I'm ok with merging this as is and adding that later (but please add a TODO/open an issue)

File: torch/csrc/autograd/python_engine.cpp [hunk 8]
```diff
@@ -201,9 +202,11 @@ PyObject *THPEngine_run_backward(THPEngine *self, PyObject *args, PyObject *kwar
       if (is_leaf) {
           grad_fn = input_var->cdata->grad_accumulator.lock();
       }
-      THPUtils_assert(grad_fn, "One of the differentiated Variables appears to not have "
-          "been used in the graph");
-      THPUtils_assert(grad_fn->is_executable, "One of the differentiated Variables does "
+      if (input_var->cdata->requires_grad) {
+        THPUtils_assert(grad_fn, "One of the differentiated Variables appears to not have "
+            "been used in the graph");
+      }
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/python_engine.cpp]: What leaf Variable case are you thinking of? Here are the error messages when `input_var` is a leaf:

1) Leaf does not require grad, used in graph:
`"One of the differentiated Variables does not require grad"`
(unchanged)

2) Leaf does not require grad, unused in graph:
`"One of the differentiated Variables does not require grad"`
Previously:
`"One of the differentiated Variables appears to not have been used in the graph"`

3) Leaf requires grad and never used:
`"One of the differentiated Variables appears to not have been used in the graph"`
(unchanged)

4) Leaf requires grad, used for different operations, but not in this graph:
`"differentiated input is unreachable"`
(unchanged)

The error message for (2) seems fine to me, since input_var always needs to have `requires_grad=True`.

**Comment by colesbury** [comment]
@apaszke, can you look this over again?

File: torch/csrc/autograd/functions/utils.cpp [hunk 9]
```diff
@@ -14,9 +14,7 @@ variable_list wrap_outputs(const variable_list& inputs, tensor_list&& outputs,
   if (flags.is_volatile) {
     for (auto& output : outputs) {
       if (output.defined()) {
-        auto pImpl = new VariableTensor(output);
-        pImpl->is_volatile = true;
-        result.emplace_back(pImpl, false);
+        result.emplace_back(new VariableTensor(output, false, true), false);
```
Comment (apaszke) [inline torch/csrc/autograd/functions/utils.cpp]: I don't think this fixes the problem I mentioned. You only got rid of this temporary assignment, but this `VariableTensor` can still leak if the vector fails to resize (it has to increase its size before it even creates the VariableTensor. I think the right way to do it would be to construct a `Variable` here, so that it takes ownership of the `VariableTensor`, and only then will be moved into the vector

File: torch/csrc/autograd/input_buffer.cpp [hunk 10]
```diff
@@ -9,7 +9,7 @@ InputBuffer::InputBuffer(size_t size)
   : buffer(size)
   {}
 
-void InputBuffer::add(size_t pos, const Variable& var) {
+void InputBuffer::add(size_t pos, Variable var) {
```
Comment (apaszke) [inline torch/csrc/autograd/input_buffer.cpp]: nit: it used to be an r-value reference. it seemed to be better than a regular value

Comment (colesbury) [reply torch/csrc/autograd/input_buffer.cpp]: We were trying to signify that InputBuffer takes ownership of `var`, but the standard way to do that with smart shared pointers is just to take it by value:

See (e) in: https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/

**Review by apaszke — APPROVED** [review]
